### ISSUE TITLE
Incorrect fix suggestion for `manual_retain`

### ISSUE THREAD

**Issue by firefighterduck** [rust-lang/rust-clippy#10393]
### Summary

The `manual_retain` lint can make clippy suggest incorrect code in the case of a vector with tuples.
To be more precise, it suggests a call to `retain` with a closure that doesn't take a tuple but two elements and introduces a wrong mutable reference to one of the elements in the tuple. Both problems lead to the code not compiling.

This works in both the current stable and nightly and is easily observable on the playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7c3b850738f1246bfb92a23cffe8caf1

### Reproducer

I tried this code:

```rust
let mut tuples = vec![(true, -2), (false, 3)];
tuples = tuples.into_iter().filter(|(_, n)| *n > 0).collect();
```

I expected to see this happen:
When running clippy with `cargo clippy`, it should warn me about the manual retaining and tell me to use `tuples.retain(|(_, n)| *n > 0)` instead.

Instead, this happened:
Clippy suggested `tuples.retain(|_, &mut n| *n > 0)` which is wrong as the closure doesn't expect a single tuple and tries to take the second argument by mutable reference even though `retain` only allows for shared references as arguments to the given function.

### Version

```text
rustc 1.67.1 (d5a82bbd2 2023-02-07)
binary: rustc
commit-hash: d5a82bbd26e1ad8b7401f6a718a9c57c96905483
commit-date: 2023-02-07
host: x86_64-unknown-linux-gnu
release: 1.67.1
LLVM version: 15.0.6
```

### Additional Labels

@rustbot label +I-suggestion-causes-error
